psl-core/src/main/java/org/linqs/psl/application/inference/online/OnlineServer.java

```diff
@@ -0,0 +1,286 @@
+
+/**
+ * A class for listening for new client connections, queueing actions from OnlineClients, and sending OnlineResponses.
+ */
+public class OnlineServer {
+    private static final Logger log = LoggerFactory.getLogger(OnlineServer.class);
+
+    private boolean listening;
+    private ServerConnectionThread serverThread;
+    private Set<ClientConnectionThread> clientConnectionThreads;
+    private BlockingQueue<OnlineMessage> queue;
+    private ConcurrentMap<UUID, ClientConnectionThread> messageIDConnectionMap;
+
+    public OnlineServer() {
+        listening = false;
+        serverThread = new ServerConnectionThread();
+        queue = new LinkedBlockingQueue<OnlineMessage>();
+        messageIDConnectionMap = new ConcurrentHashMap<UUID, ClientConnectionThread>();
+        clientConnectionThreads = new HashSet<ClientConnectionThread>();
+    }
+
+    /**
+     * Start up the server and listen for new connections on the configured port.
+     * This method does not block. A new thread is started to wait for connections.
+     */
+    public void start() {
+        listening = true;
+        serverThread.start();
+    }
+
+    /**
+     * Get the next action from the client.
+     * This method will block until an action is available to take from the queue.
+     */
+    public OnlineMessage getAction() {
+        OnlineMessage nextAction = null;
+
+        while (nextAction == null) {
+            try {
+                nextAction = queue.take();
+            } catch (InterruptedException ex) {
+                log.warn("Interrupted while taking an online action from the queue.", ex);
+                return null;
+            }
+
+            if (nextAction instanceof Exit) {
+                onActionExecution(nextAction, new ActionStatus(nextAction, true, "Session Closed."));
+                nextAction = null;
+            }
+        }
+
+        return nextAction;
+    }
+
+    public void onActionExecution(OnlineMessage action, OnlineResponse onlineResponse) {
+        ClientConnectionThread clientConnectionThread = messageIDConnectionMap.get(action.getIdentifier());
+        ObjectOutputStream outputStream = clientConnectionThread.outputStream;
+
+        try {
+            outputStream.writeObject(onlineResponse);
+        } catch (IOException ex) {
+            log.warn(String.format("Failed to send client onlineResponse: %s", onlineResponse));
+        }
+
+        if (action instanceof Exit || action instanceof Stop) {
+            closeClient(clientConnectionThread);
+        }
+
+        if (onlineResponse instanceof ActionStatus) {
+            messageIDConnectionMap.remove(action.getIdentifier());
+        }
+    }
+
+    public void closeClient(ClientConnectionThread clientConnectionThread) {
+        clientConnectionThread.close();
+        clientConnectionThreads.remove(clientConnectionThread);
+    }
+
+    public void addClient(ClientConnectionThread clientConnectionThread) {
```

Comment:
Race condition, this set is accessed by the different threads.
I would use [Collections.synchronizedSet()](https://docs.oracle.com/javase/7/docs/api/java/util/Collections.html#synchronizedSet(java.util.Set)) for this set.